### PR TITLE
Move caching to Contour.Caching

### PR DESCRIPTION
**Pull request by andrey-prokopyev**
Added extension features needed for moving caching to Contour.Caching:
- config sections extensions
- decorating outgoing message filters
- removing code responsible for caching

### CODE REVIEW

**Comment by semion-akimtsev** [comment]
@andrey-prokopyev ,

could you please provide description of the intent and changes of this pull request?

File: Sources/Contour.Configuration/Configurator/CallbackEndpointElement.cs [hunk 1]
```diff
@@ -5,7 +5,7 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The callback endpoint element.
     /// </summary>
-    internal class CallbackEndpointElement : ConfigurationElement
+    public class CallbackEndpointElement : ConfigurationElement
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/CallbackEndpointElement.cs:8]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/EndpointCollection.cs [hunk 2]
```diff
@@ -6,7 +6,7 @@ namespace Contour.Configuration.Configurator
     /// The endpoint collection.
     /// </summary>
     [ConfigurationCollection(typeof(EndpointElement), AddItemName = "endpoint")]
-    internal class EndpointCollection : ConfigurationElementCollection
+    public class EndpointCollection : ConfigurationElementCollection
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/EndpointCollection.cs:9]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/EndpointElement.cs [hunk 3]
```diff
@@ -6,15 +6,9 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The endpoint element.
     /// </summary>
-    internal class EndpointElement : ConfigurationElement
+    public class EndpointElement : ExtensibleConfigurationElement
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/EndpointElement.cs:9]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/EndpointsSection.cs [hunk 4]
```diff
@@ -5,7 +5,7 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The endpoints section.
     /// </summary>
-    internal class EndpointsSection : ConfigurationSection
+    public class EndpointsSection : ConfigurationSection
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/EndpointsSection.cs:8]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/IncomingCollection.cs [hunk 5]
```diff
@@ -6,7 +6,7 @@ namespace Contour.Configuration.Configurator
     /// The incoming collection.
     /// </summary>
     [ConfigurationCollection(typeof(IncomingElement), AddItemName = "on")]
-    internal class IncomingCollection : ConfigurationElementCollection
+    public class IncomingCollection : ConfigurationElementCollection
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/IncomingCollection.cs:9]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/IncomingElement.cs [hunk 6]
```diff
@@ -5,7 +5,7 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The incoming element.
     /// </summary>
-    internal class IncomingElement : MessageElement
+    public class IncomingElement : MessageElement
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/IncomingElement.cs:8]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/LambdaDependencyResolver.cs [hunk 7]
```diff
@@ -16,7 +16,7 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The lambda dependency resolver.
     /// </summary>
-    internal class LambdaDependencyResolver : IDependencyResolver
+    public class LambdaDependencyResolver : IDependencyResolver
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/LambdaDependencyResolver.cs:19]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/MessageElement.cs [hunk 8]
```diff
@@ -1,11 +1,11 @@
...
+{
+    using System.Configuration;
 
-namespace Contour.Configuration.Configurator
-{
     /// <summary>
     /// The message element.
     /// </summary>
-    internal class MessageElement : ConfigurationElement
+    public class MessageElement : ExtensibleConfigurationElement
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/MessageElement.cs:8]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/OutgoingCollection.cs [hunk 9]
```diff
@@ -6,7 +6,7 @@ namespace Contour.Configuration.Configurator
     /// The outgoing collection.
     /// </summary>
     [ConfigurationCollection(typeof(OutgoingElement), AddItemName = "route")]
-    internal class OutgoingCollection : ConfigurationElementCollection
+    public class OutgoingCollection : ConfigurationElementCollection
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/OutgoingCollection.cs:9]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/OutgoingElement.cs [hunk 10]
```diff
@@ -6,7 +6,7 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The outgoing element.
     /// </summary>
-    internal class OutgoingElement : MessageElement
+    public class OutgoingElement : MessageElement
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/OutgoingElement.cs:9]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/ValidatorCollection.cs [hunk 11]
```diff
@@ -6,7 +6,7 @@ namespace Contour.Configuration.Configurator
     /// The validator collection.
     /// </summary>
     [ConfigurationCollection(typeof(ValidatorElement), AddItemName = "add")]
-    internal class ValidatorCollection : ConfigurationElementCollection
+    public class ValidatorCollection : ConfigurationElementCollection
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/ValidatorCollection.cs:9]: Why should this type be public?

File: Sources/Contour.Configuration/Configurator/ValidatorElement.cs [hunk 12]
```diff
@@ -5,7 +5,7 @@ namespace Contour.Configuration.Configurator
     /// <summary>
     /// The validator element.
     /// </summary>
-    internal class ValidatorElement : ConfigurationElement
+    public class ValidatorElement : ConfigurationElement
```
Comment (semion-akimtsev) [inline Sources/Contour.Configuration/Configurator/ValidatorElement.cs:8]: Why should this type be public?

**Comment by andrey-prokopyev** [comment]
@semion-akimtsev, done

File: Sources/Contour/Configuration/BusConfiguration.cs [hunk 13]
```diff
@@ -342,6 +331,11 @@ public void RegisterFilter(IMessageExchangeFilter filter)
             this.filters.Add(filter);
         }
 
+        public void RegisterDecoratorOf<T>(IMessageExchangeFilterDecorator decorator)
+        {
+            this.FilterDecorators[typeof(T)] = decorator;
```
Comment (semion-akimtsev) [inline Sources/Contour/Configuration/BusConfiguration.cs]: The type of what entity should be used as T?

File: Sources/Contour/Contour.csproj [hunk 14]
```diff
@@ -41,15 +41,12 @@
     <Reference Include="System.Core" />
     <Reference Include="System.Runtime.Caching" />
     <Reference Include="Microsoft.CSharp" />
+    <Reference Include="System.Xml" />
```
Comment (semion-akimtsev) [inline Sources/Contour/Contour.csproj:44]: Why is this dependency needed?

File: Sources/Contour/Filters/IMessageExchangeFilterDecorator.cs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Contour.Filters
+{
+    using System.Threading.Tasks;
```
Comment (semion-akimtsev) [inline Sources/Contour/Filters/IMessageExchangeFilterDecorator.cs]: Move the using directive outside the namespace scope.

File: Sources/Contour/Filters/IMessageExchangeFilterDecorator.cs [hunk 16]
```diff
@@ -0,0 +1,9 @@
...
+
+    public interface IMessageExchangeFilterDecorator
```
Comment (semion-akimtsev) [inline Sources/Contour/Filters/IMessageExchangeFilterDecorator.cs]: What types implement this interface?

File: Sources/Contour/Filters/MessageExchangeFilterInvoker.cs [hunk 17]
```diff
@@ -24,6 +29,13 @@ public MessageExchangeFilterInvoker(IEnumerable<IMessageExchangeFilter> filters)
             this.filterEnumerator = filters.Reverse().
                 GetEnumerator();
         }
+
+        public MessageExchangeFilterInvoker(IEnumerable<IMessageExchangeFilter> filters, IDictionary<Type, IMessageExchangeFilterDecorator> filterDecorators)
+            : this(filters)
+        {
+            this.filterDecorators = filterDecorators;
```
Comment (semion-akimtsev) [inline Sources/Contour/Filters/MessageExchangeFilterInvoker.cs]: Shouldn't the filterDecorators collection be checked for null here?

File: Sources/Contour/Serialization/JsonNetPayloadConverter.cs [hunk 18]
```diff
@@ -8,7 +8,7 @@
     /// <summary>
     /// The json net payload converter.
     /// </summary>
-    internal class JsonNetPayloadConverter : IPayloadConverter
+    public class JsonNetPayloadConverter : IPayloadConverter
```
Comment (semion-akimtsev) [inline Sources/Contour/Serialization/JsonNetPayloadConverter.cs:11]: Why should this type be public?

**Comment by semion-akimtsev** [comment]
Where Contour.Caching and corresponding tests are located?

**Review by semion-akimtsev — CHANGES_REQUESTED** [review]
Please make this fork merge-able with the silver-lizard branch.

**Comment by andrey-prokopyev** [comment]
@semion-akimtsev, configuration types must be public because of their extensibility. Side projects should be able to read ExtensionProperties and ExtensionConfigurations. Other internal config elements became public because of their visibility as types of public properties.

**Comment by andrey-prokopyev** [comment]
@semion-akimtsev, LambdaDependencyResolver became public beacuse of it's use in Contour.Caching

**Comment by andrey-prokopyev** [comment]
@semion-akimtsev, System.Xml is used in ExtensibleConfigurationElement

**Comment by andrey-prokopyev** [comment]
@semion-akimtsev, JsonNetPayloadConverter is used in message hasher of Contour.Caching

**Comment by andrey-prokopyev** [comment]
@semion-akimtsev , Contour.Caching is located at our internal new.git.dev

File: Sources/Contour/Filters/MessageExchangeFilterInvoker.cs [hunk 19]
```diff
@@ -25,15 +25,15 @@ public class MessageExchangeFilterInvoker
...
         {
-            this.filterEnumerator = filters.Reverse().
-                GetEnumerator();
         }
 
         public MessageExchangeFilterInvoker(IEnumerable<IMessageExchangeFilter> filters, IDictionary<Type, IMessageExchangeFilterDecorator> filterDecorators)
-            : this(filters)
         {
-            this.filterDecorators = filterDecorators;
+            this.filterEnumerator = filters.Reverse().
```
Comment (semion-akimtsev) [inline Sources/Contour/Filters/MessageExchangeFilterInvoker.cs]: The IEnumerable collection does not guarantee the order of its elements. What should the correct order be here?

File: Sources/Contour/Filters/IMessageExchangeFilterDecorator.cs [hunk 20]
```diff
@@ -1,7 +1,7 @@
-﻿namespace Contour.Filters
-{
-    using System.Threading.Tasks;
+﻿using System.Threading.Tasks;
 
+namespace Contour.Filters
+{
     public interface IMessageExchangeFilterDecorator
```
Comment (semion-akimtsev) [inline Sources/Contour/Filters/IMessageExchangeFilterDecorator.cs]: A type which should be used as a decorator must implement an interface of the decorated object per definition. See http://www.oodesign.com/decorator-pattern.html for detailed explanation. The purpose of this is to hide the actual difference between the original object and the augmented one from the consumers of the interface. Why does IMessageExchangeFilterDecorator not do it?

**Review by semion-akimtsev — COMMENTED** [review]
Please take a look at the comments below.... or above :)

**Comment by semion-akimtsev** [comment]
@andrey-prokopyev , @mzabolotko ,

I see that we have a couple of changes (types accessibility levels etc.) in Contour which should affect the behavior of some other library which is currently "private". These changes cannot be validated now and we cant justify them because the "other" library is not part of the solution. I suggest that we move Contour.Caching to this repository (not assembly) and make it build-able because despite it should be detached from Contour itself it is still a part of it as a framework. Or rename the library to something_cool_not_referencing_contour.Caching and make it dependent on the branch of Contour project.

**Comment by andrey-prokopyev** [comment]
@mzabolotko, so Contour.Caching should be separate repository or separate project in Contour repo?

**Pull request closed without merging**